federation/pkg/federatedtypes/hpa_test.go

```diff
@@ -0,0 +1,262 @@
+			fedHpa: newHpaWithReplicas(NewInt32(4), NewInt32(70), 10),
+			localHpas: func() map[string]pkgruntime.Object {
+				hpa1 := newHpaWithReplicas(NewInt32(3), NewInt32(70), 6)
+				hpa1 = updateHpaStatus(hpa1, NewInt32(50), 4, 4, true)
+				hpa2 := newHpaWithReplicas(NewInt32(1), NewInt32(70), 4)
+				hpa2 = updateHpaStatus(hpa2, NewInt32(50), 3, 3, true)
+				hpas := make(map[string]pkgruntime.Object)
+				hpas["c1"] = hpa1
+				hpas["c2"] = hpa2
+				return hpas
+			}(),
+			expectedReplicas: map[string]*replicas{
+				"c1": {
+					min: int32(3),
+					max: int32(6),
+				},
+				"c2": {
+					min: int32(1),
+					max: int32(4),
+				},
+			},
+		},
+		"Increasing replicas on fed object increases the same on clusters": {
+			// Existing total of local min, max = 1+1, 5+5 decreasing to below
+			fedHpa: newHpaWithReplicas(NewInt32(4), NewInt32(70), 14),
+			localHpas: func() map[string]pkgruntime.Object {
+				// does not matter if scaleability is true
+				hpas := make(map[string]pkgruntime.Object)
+				hpas["c1"] = newHpaWithReplicas(NewInt32(1), NewInt32(70), 5)
+				hpas["c2"] = newHpaWithReplicas(NewInt32(1), NewInt32(70), 5)
+				return hpas
+			}(),
+			// We dont know which cluster gets how many, but the resultant total should match
+		},
+		"Decreasing replicas on fed object decreases the same on clusters": {
+			// Existing total of local min, max = 2+2, 8+8 decreasing to below
+			fedHpa: newHpaWithReplicas(NewInt32(3), NewInt32(70), 8),
+			localHpas: func() map[string]pkgruntime.Object {
+				// does not matter if scaleability is true
+				hpas := make(map[string]pkgruntime.Object)
+				hpas["c1"] = newHpaWithReplicas(NewInt32(2), NewInt32(70), 8)
+				hpas["c2"] = newHpaWithReplicas(NewInt32(2), NewInt32(70), 8)
+				return hpas
+			}(),
+			// We dont know which cluster gets how many, but the resultant total should match
+		},
+	}
+
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			if testCase.fedHpa == nil {
+				testCase.fedHpa = defaultFedHpa
+			}
+			scheduledState := getHpaScheduleState(testCase.fedHpa, testCase.localHpas)
+			checkClusterConditions(t, testCase.fedHpa, scheduledState)
+			if testCase.expectedReplicas != nil {
+				for cluster, replicas := range testCase.expectedReplicas {
+					scheduledReplicas := scheduledState[cluster]
+					assert.Equal(t, replicas.min, scheduledReplicas.min)
+					assert.Equal(t, replicas.max, scheduledReplicas.max)
+				}
+			}
+		})
+	}
+}
+
+func updateHpaStatus(hpa *autoscalingv1.HorizontalPodAutoscaler, currentUtilisation *int32, current, desired int32, scaleable bool) *autoscalingv1.HorizontalPodAutoscaler {
+	hpa.Status.CurrentReplicas = current
+	hpa.Status.DesiredReplicas = desired
+	hpa.Status.CurrentCPUUtilizationPercentage = currentUtilisation
+	now := metav1.Now()
+	scaledTime := now
+	if scaleable {
+		// definitely more then 5 minutes ago
+		scaledTime = metav1.NewTime(now.Time.Add(-6 * time.Minute))
```

Comment:
Actually, I did make this particular time configurable on a separate PR https://github.com/kubernetes/kubernetes/pull/49583. I have put good amount of comments there to explain the same: https://github.com/kubernetes/kubernetes/pull/49583/files#diff-e303386d7ba192e209b310d96206a157R75.